### PR TITLE
Fetch title ID from Homebrew Launcher Mod

### PR DESCRIPTION
**Pull request by suloku**
My current homebrew launcher mod stores the u64 tid in /svdt/tid.bin when launching an app with title launcher selection who's smdh name is "svdt".
This modification allows to start svdt with the selected title.
Why not using argv[1]? Because even though it works fine with netloading, I can't manage to set argv[1] even with the netloader code.

### CODE REVIEW

**Comment by meladroit** [comment]
So if I understand this correctly (full disclaimer: I haven't compiled/tested svdt with this patch yet), the intent of this commit is to make it so that svdt will automatically fetch the tid from your modded 3ds_hb_menu, but will still give you the option of manually setting the title? Seems fine to me.

However, if tid.bin is somehow corrupted so that it is still readable but none of the title IDs in our list match its contents, then it leaves titleTitle unchanged. I don't think it's uninitialised (probably a timestamp?) but I don't know if it's desired behaviour, really. Couldn't you do nthTitleInList(titleTitle_set,[...]) first, go through the for loop if tid2 != 0, and then if the loop finishes without a match, go back to nthTitleInList(titleTitle_set,[...]) as a default?

(Actually, you shouldn't even need to do nthTitleInList in a for loop—that's quite overkill if you just want to look through the title IDs. You just need the tid, so go through the linked list without using wrapper functions, look for a tid match, and if tid2 == tid then titleTitle_set = i and break, otherwise titleTitle_set = 0, and then call nthTitleInList. You shouldn't even need the for loop, really. Just do a loop with while(thisTitle) or something (check to see current title isn't NULL, which indicates the end of the linked list), continually increment i but keep it separate from titleTitle_set until you have a match, and you're clear.)

Plus, I think you'd want to do titleTitle_set = i before breaking anyway, if your intention is to let the user still scroll through available titles after fetching the automatic tid. Having titleTitle_set wrong won't affect anything technically (everything will right itself once the user starts pressing left or right to scroll through the list), but it'd be nice from a user standpoint.

**Comment by suloku** [comment]
Ok, I've implemented your "suggestions", which were almost the code itself. Tested and working fine, also tested corrupt file.

Also, may I suggest changing the hardcoded "/3ds/svdt/" paths (for flags and asr.dat) to either /svdt or to use relative paths? The homebrew launcher (vanilla) sets the current directory to that of the booted app before launching, so "asr.dat" and the flags should work fine without the /3ds/svdt/ part of the path.
This helps having multiple versions of svdt too, for example one could edit an xml file to use game XXX title id, so the title selection screen is bypassed.
Having hardcoded paths for the 3dsx file limits the possibilities and there shouldn't be any reason to use them since everybody boots from the homebrew lacuncher.

**Comment by meladroit** [comment]
You're right—you'd have to use 3ds_hb_menu to get svdt working properly anyway. I think I would favour relative paths in that case, because of the use case you mention of having copies of svdt with targets selectable="false". (Plus, /svdt/ was always meant to be solely for save data backups.)

If you'd like to implement relative paths before I merge this pull request, go ahead. Despite my very strong 'suggestions', I'm quite badly burnt out on writing code at the moment.

**Comment by suloku** [comment]
Ok, leave it to me, not that many people use my mod anyway (I think). Someone suggested a different function to pass argv[1], so I'll test and see if I can make it work trough arguments.

**Comment by suloku** [comment]
Seems I was too eager to commit, even though creating a file with relative paths works fine, reading seems to be different? I can't make it to work with relative paths.
Also, oot3dhax installer uses relative paths, but fails when not run from the /3ds/ folder. This is frustrating, why does creating a file work with relative path but reading doesn't? Doesn't make any sense

**Comment by suloku** [comment]
Ok, I had to create a file_exist() function with standard C functions due to your doesFileNotExist() function not working with relative paths.

Tested and everything works fine right now. I hope you are ok with the changes.

Also, smash bros files are inside a /save_data folder/, you might not know due to not owning the game and reading the information from my post in gbatemp.

**Comment by suloku** [comment]
Ok, currently everything works fine, except for the fact that when rewriting secure value for the smash bros Demo I get a "Error rewriting secure value. [result code ffffffff]".

The secure value file is opened with fopen(), so there should be no problem there. Did I not understand how ars.dat is supposed to work?

**Comment by meladroit** [comment]
Sorry I've been absent. So the reason doesFileNotExist doesn't work with relative paths is that it calls FS:USER functions, which essentially always require an absolute path. But you seem to have addressed that by now.

No, asr.dat is supposed to work exactly the way you've written it. However, it looks like you haven't revised secure_values.c to use relative paths as well. Do that, throw your name in the instructions or title bar somewhere (update the version number to v0.10.1 or something, too), and I think we should be able to call it a day.

Thanks for all the work!

**Comment by meladroit** [comment]
Wait, I stand corrected—you are using relative paths in secure_values.c. I was looking at the wrong branch. readBytesFromSaveFile and writeBytesToSaveFile are not using relative paths, but neither is asr.dat.

Still, it's either failing at !checkSecureConfig() or !secureValueSet if it's returning (u32)(-1). One thing worth noting is that secure_values.c seems to use if(checkSecureConfig()) for getSecureValue2 but if(!checkSecureConfig()) for writeSecureValue2. It's perfectly possible one of those is wrong. In fact, I'm almost certain it should be if(checkSecureConfig()) return -1; for both of them, since we want it to fail out if asr.dat doesn't exist. (It's quite plausible that I never tested writing to save data with asr.dat enabled.)

**Comment by suloku** [comment]
OK, everything is working right now. I forgot to change a call to doesFileNotExist(), but the code was missing a ! in the checkSecureConfig() check, which is what was making it fail.

I've changed the logit so checkSecureConfig()returns 1, as file_exist() returns 1 if the file exist (which is what we want to check), and not if the file doesn't exist.
I think this is more clear than negation, but I could change file_exist() return values so it works like doesFileNotExist() if you prefer that.

To sum up: right now everything is all right.

ps: I want to check tomorrow if Pokemon ORAS demo has a secure value, so maybe you want to wait for a release with asr.dat updated with that.

**Comment by meladroit** [comment]
Sure, I'll wait for that update. Thanks again!

**Comment by suloku** [comment]
Ok, I've located secure value at 0x5800, but for some reason svdt refuses to write to that save while save_manager works fine.

Delete: console says everything is fine, but save is not deleted.
Overwrite: fatal error.

Maybe it is because the file is named "main" and it isn't either X/Y or OR. If it is that, I'll just have to remove those checks when asr is disabled, I haven't really checked the code yet so this is just guessing.

**Comment by meladroit** [comment]
You mean disabling anti-ASR at the startup prompt somehow doesn't disable those checks? It really should, since it sets whichSecureGame back to SECURE_UNKNOWN, which should mean svdt never even thinks about ASR ... Plus, if I've done this right, secureGameFromProductCode should override any filesystem-based guesses ...

**Comment by suloku** [comment]
No, I mean that retrieving asr from filesystem is conflicting with pokemon ORAS demo since it also uses a "main" file. I haven't really checked yet.

Don't worry, I'll sort it out later.

**Comment by meladroit** [comment]
Yeah, that makes sense. The best option might be to remove the Pokémon 'main' game filesystem checks (I think they all work fine with svdt anyway). So I'll merge the current state of things for now and do that, and you can submit another pull request if necessary.

**Comment by suloku** [comment]
Tested and the conflict is gone, everything working, and the secure value of the Demo really was at 0x5800.

So, I can think of different solutions for filesystem detection:
1.- Change detection to filesize: I'm pretty sure XY, ORAS and OR Demo savegames have a constant sized savegame, but a japan confirmation would be nice.
2.- Add the offset for OR Demo to the checks, but the problem will arise again if a new pokemon game that uses the savefile /main comes out (and I think one is commin out on christmas)
3.- Any of the latest, but add user confirmation: "Secure value inferred from filesyste, Is this a XXX savegame? A/B". The user is in charge of confirming the detection worked.

I'm for number 3, as this would also solve any other possible conflicts (rumble blast may be one)

**Comment by meladroit** [comment]
Well, my thought is that we should absolutely never do anything in interactive mode based on filesystem-based guesses, which rules out 3. One way or another, if we get to the point where we can actually display text to the user, we have an actual title ID we can use to unambiguously know which game we're dealing with.

Until there's a Pokémon main game on 3DS that actually doesn't let svdt start up properly (and to my knowledge there is no such game), I'd lean towards holding off on filesystem-based guesses and force the user to manually enable anti-ASR for these games.
